codegen/value.go

```diff
@@ -52,9 +54,71 @@ func NewRegister(ctx context.Context) Register {
 	}
 }
 
-func (r *register) Set(iface interface{}) (err error) {
-	r.Value, err = r.ctor(iface)
-	return err
+func (r *register) Set(iface interface{}) error {
+	return r.SetAsync(func(Value) (Value, error) {
+		return r.ctor(iface)
+	})
+}
+
+func (r *register) SetAsync(f func(Value) (Value, error)) error {
+	var prev chan Value
+	if len(r.rets) == 0 {
+		prev = make(chan Value, 1)
+		prev <- r.Value
```

Comment:
Trying to understand this pattern. You create a buffered channel, write to it, and immediately read the value in the goroutine below?

I think some comments would help here.

I'm also not sure I understand the stack of `rets` and why it makes sense for `value()` to pop from the stack.

Reply:
So I'm essentially creating a chain of goroutines piped together with channels. For the zero case, I create a buffered channel to bootstrap the chain.

You're right, I don't actually need the stack, I simplified it to just keeping track of the last channel.